packages/flutter/lib/src/material/button_style.dart

```diff
@@ -0,0 +1,513 @@
+      overlayColor: _lerpColors(a?.overlayColor, b?.overlayColor, t),
+      shadowColor: _lerpColors(a?.shadowColor, b?.shadowColor, t),
+      elevation: _lerpDoubles(a?.elevation, b?.elevation, t),
+      padding: _lerpInsets(a?.padding, b?.padding, t),
+      minimumSize: _lerpSizes(a?.minimumSize, b?.minimumSize, t),
+      side: _lerpSides(a?.side, b?.side, t),
+      shape: _lerpShapes(a?.shape, b?.shape, t),
+      mouseCursor: t < 0.5 ? a.mouseCursor : b.mouseCursor,
+      visualDensity: t < 0.5 ? a.visualDensity : b.visualDensity,
+      tapTargetSize: t < 0.5 ? a.tapTargetSize : b.tapTargetSize,
+      animationDuration: t < 0.5 ? a.animationDuration : b.animationDuration,
+      enableFeedback: t < 0.5 ? a.enableFeedback : b.enableFeedback,
+    );
+  }
+
+  static MaterialStateProperty<TextStyle> _lerpTextStyles(MaterialStateProperty<TextStyle> a, MaterialStateProperty<TextStyle> b, double t) {
+    if (a == null && b == null)
+      return null;
+    return _LerpTextStyles(a, b, t);
+  }
+
+  static MaterialStateProperty<Color> _lerpColors(MaterialStateProperty<Color> a, MaterialStateProperty<Color> b, double t) {
+    if (a == null && b == null)
+      return null;
+    return _LerpColors(a, b, t);
+  }
+
+  static MaterialStateProperty<double> _lerpDoubles(MaterialStateProperty<double> a, MaterialStateProperty<double> b, double t) {
+    if (a == null && b == null)
+      return null;
+    return _LerpDoubles(a, b, t);
+  }
+
+  static MaterialStateProperty<EdgeInsetsGeometry> _lerpInsets(MaterialStateProperty<EdgeInsetsGeometry> a, MaterialStateProperty<EdgeInsetsGeometry> b, double t) {
+    if (a == null && b == null)
+      return null;
+    return _LerpInsets(a, b, t);
+  }
+
+  static MaterialStateProperty<Size> _lerpSizes(MaterialStateProperty<Size> a, MaterialStateProperty<Size> b, double t) {
+    if (a == null && b == null)
+      return null;
+    return _LerpSizes(a, b, t);
+  }
+
+  static MaterialStateProperty<BorderSide> _lerpSides(MaterialStateProperty<BorderSide> a, MaterialStateProperty<BorderSide> b, double t) {
+    if (a == null && b == null)
+      return null;
+    return _LerpSides(a, b, t);
+  }
+
+  static MaterialStateProperty<OutlinedBorder> _lerpShapes(MaterialStateProperty<OutlinedBorder> a, MaterialStateProperty<OutlinedBorder> b, double t) {
+    if (a == null && b == null)
+      return null;
+    return _LerpShapes(a, b, t);
+  }
+}
+
+class _LerpTextStyles implements MaterialStateProperty<TextStyle> {
+  const _LerpTextStyles(this.a, this.b, this.t);
+
+  final MaterialStateProperty<TextStyle> a;
+  final MaterialStateProperty<TextStyle> b;
+  final double t;
+
+  @override
+  TextStyle resolve(Set<MaterialState> states) {
+    final TextStyle resolvedA = a?.resolve(states);
+    final TextStyle resolvedB = b?.resolve(states);
+    return TextStyle.lerp(resolvedA, resolvedB, t);
+  }
+}
+
+class _LerpColors implements MaterialStateProperty<Color> {
```

Comment:
Do you think something along these lines could work to save a few lines of code?

```dart
class _LerpProperties<T> implements MaterialStateProperty<T> {
  const _LerpProperties(this.a, this.b, this.t, this.lerpFunction);

  final MaterialStateProperty<T> a;
  final MaterialStateProperty<T> b;
  final double t;
  final T Function(T, T, double) lerpFunction;

  @override
  T resolve(Set<MaterialState> states) {
    final T resolvedA = a?.resolve(states);
    final T resolvedB = b?.resolve(states);
    return lerpFunction(resolvedA, resolvedB, t);
  }
}
```

Then the calls could look like:

```dart
static MaterialStateProperty<TextStyle> _lerpTextStyles(MaterialStateProperty<TextStyle> a, 
  MaterialStateProperty<TextStyle> b, double t) {
    if (a == null && b == null)
      return null;
    return _LerpProperties<TextStyle>(a, b, t, TextStyle.lerp);
  }

```

Reply:
And actually, if that _did_ work, then you would be able to condense some of the methods above.

Something like:

```dart
  static MaterialStateProperty<T> _lerpProperties(MaterialStateProperty<T> a, MaterialStateProperty<T> b, double t, T Function(T, T, double) lerpFunction ) {
    if (a == null && b == null)
      return null;
    return _LerpProperties(a, b, t, lerpFunction);
  }
```

And above you would do something like:

```dart
 return ButtonStyle(
      textStyle: _lerpProperties<TextStyle>(a?.textStyle, b?.textStyle, t, TextStyle.lerp),
```

Reply:
Both suggestions are nice boilerplate reducers, I've incorporated them.